### ISSUE TITLE
False positive; args[module]: unable to evaluate string as dictionary (warning)

### ISSUE THREAD

**Issue by jamesholland-uk** [ansible/ansible-lint#3128]
##### Summary
We are receiving `args[module]: unable to evaluate string as dictionary (warning)` which we believe is a false positive. We understand this rule is still `experimental` at this time, but would like to ask that it is modified to accommodate this use case before it gets promoted, especially as we are targeting `production` profile.

Users using the Palo Alto Networks collection of modules for firewalls, frequently execute tasks using multiple modules in a single playbook. To keep the firewall connectivity parameters in a DRY state, this is a common pattern:
```
.
.
.
  vars:
    device:
      ip_address: "{{ ip_address }}"
      username: "{{ username | default(omit) }}"
      password: "{{ password | default(omit) }}"
      api_key: "{{ api_key | default(omit) }}"
.
.
.
    - name: Create import rule 1
      paloaltonetworks.panos.panos_bgp_policy_rule:
        provider: "{{ device }}"
        vr_name: "default"
        name: "rule-001"
        type: "import"
        address_prefix:
          - name: "10.1.1.0/24"
        enable: true
        action: "allow"

    - name: Create import rule 2
      paloaltonetworks.panos.panos_bgp_policy_rule:
        provider: "{{ device }}"
        vr_name: "default"
        name: "rule-002"
        type: "import"
        address_prefix:
          - name: "10.2.2.0/24"
        enable: true
        action: "allow"
.
.
.
```

It seems that using the `device` var to substitute the firewall connectivity parameters causes the `args[module]` warning to trigger. We can make the warning go away with something like this where the parameters are explicitly set per task, but that makes the playbooks very un-DRY:
```
    - name: Create import rule 1
      paloaltonetworks.panos.panos_bgp_policy_rule:
        provider:
          ip_address: "192.168.1.1"
          username: "admin"
          password: "something"
.
.
.
```

##### Issue Type

- Bug Report

##### OS / ENVIRONMENT
```console (paste below)
ansible-lint --version
ansible-lint 6.13.1 using ansible 2.14.3
```

- ansible installation method: pip
- ansible-lint installation method: pip
```
ansible [core 2.14.3]
  config file = /Users/jholland/code/github/ansible-playbooks/ansible.cfg
  configured module search path = ['/Users/jholland/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/jholland/.pyenv/versions/3.10.4/lib/python3.10/site-packages/ansible
  ansible collection location = /Users/jholland/.ansible/collections:/usr/share/ansible/collections
  executable location = /Users/jholland/.pyenv/versions/3.10.4/bin/ansible
  python version = 3.10.4 (main, May 24 2022, 14:08:56) [Clang 13.1.6 (clang-1316.0.21.2.5)] (/Users/jholland/.pyenv/versions/3.10.4/bin/python)
  jinja version = 3.1.2
  libyaml = False
```
- Macbook: M1 Pro, Ventura 13.2
 
##### STEPS TO REPRODUCE
```console (paste below)
ansible-lint -p upgrade_single.yml
```
(Note that `upgrade_single.yml` refers to [this playbook](https://github.com/PaloAltoNetworks/ansible-playbooks/blob/master/upgrade_single.yml). Also [this is a MCVE playbook](https://github.com/jamesholland-uk/playground/blob/main/pan-os-ansible/mcve.yml))

##### Desired Behavior
ansible-lint produces no instances of `args[module]: unable to evaluate string as dictionary (warning)`

##### Actual Behavior
ansible-lint produces three instances of `args[module]: unable to evaluate string as dictionary (warning)` are returned which are not desired.
(Note that other ansible-lint warnings for this playbook for "play name" and "jinja spacing" are known and understood, they are left there at present merely as a control point.)

Execution with extra verbosity:

```
$ ansible-lint -p upgrade_single.yml -vvv
DEBUG    Logging initialized to level 10
DEBUG    Options: Namespace(cache_dir='/Users/jholland/.cache/ansible-compat/da1f3a', colored=True, configured=True, cwd=PosixPath('/Users/jholland/code/github/ansible-playbooks'), display_relative_path=True, exclude_paths=['.cache', '.git', '.hg', '.svn', '.tox'], format=None, lintables=['upgrade_single.yml'], list_rules=False, list_tags=False, write_list=None, parseable=True, quiet=0, rulesdirs=['/Users/jholland/.pyenv/versions/3.10.4/lib/python3.10/site-packages/ansiblelint/rules'], skip_list=[], tags=[], verbosity=3, warn_list=['experimental', 'jinja[spacing]'], kinds=[{'jinja2': '**/*.j2'}, {'jinja2': '**/*.j2.*'}, {'yaml': '.github/**/*.{yaml,yml}'}, {'text': '**/templates/**/*.*'}, {'execution-environment': '**/execution-environment.yml'}, {'ansible-lint-config': '**/.ansible-lint'}, {'ansible-lint-config': '**/.config/ansible-lint.yml'}, {'ansible-navigator-config': '**/ansible-navigator.{yaml,yml}'}, {'inventory': '**/inventory/**.{yaml,yml}'}, {'requirements': '**/meta/requirements.{yaml,yml}'}, {'galaxy': '**/galaxy.yml'}, {'reno': '**/releasenotes/*/*.{yaml,yml}'}, {'tasks': '**/tasks/**/*.{yaml,yml}'}, {'rulebook': '**/rulebooks/*.{yml,yaml'}, {'playbook': '**/playbooks/*.{yml,yaml}'}, {'playbook': '**/*playbook*.{yml,yaml}'}, {'role': '**/roles/*/'}, {'handlers': '**/handlers/*.{yaml,yml}'}, {'vars': '**/{host_vars,group_vars,vars,defaults}/**/*.{yaml,yml}'}, {'test-meta': '**/tests/integration/targets/*/meta/main.{yaml,yml}'}, {'meta': '**/meta/main.{yaml,yml}'}, {'meta-runtime': '**/meta/runtime.{yaml,yml}'}, {'arg_specs': '**/meta/argument_specs.{yaml,yml}'}, {'yaml': '.config/molecule/config.{yaml,yml}'}, {'requirements': '**/molecule/*/{collections,requirements}.{yaml,yml}'}, {'yaml': '**/molecule/*/{base,molecule}.{yaml,yml}'}, {'requirements': '**/requirements.{yaml,yml}'}, {'playbook': '**/molecule/*/*.{yaml,yml}'}, {'yaml': '**/{.ansible-lint,.yamllint}'}, {'changelog': '**/changelogs/changelog.yaml'}, {'yaml': '**/*.{yaml,yml}'}, {'yaml': '**/.*.{yaml,yml}'}], mock_filters=[], mock_modules=[], mock_roles=[], loop_var_prefix=None, only_builtins_allow_collections=[], only_builtins_allow_modules=[], var_naming_pattern=None, offline=None, project_dir='.', extra_vars=None, enable_list=[], skip_action_validation=True, strict=False, rules={}, profile=None, task_name_prefix='{stem} | ', sarif_file=None, list_profiles=False, progressive=False, rulesdir=[], use_default_rules=False, generate_ignore=False, config_file=None, version=False, cache_dir_lock=<filelock._unix.UnixFileLock object at 0x102a317b0>)
DEBUG    /Users/jholland/code/github/ansible-playbooks
DEBUG    Effective yamllint rules used: {'braces': {'level': 'error', 'forbid': False, 'min-spaces-inside': 0, 'max-spaces-inside': 1, 'min-spaces-inside-empty': -1, 'max-spaces-inside-empty': -1}, 'brackets': {'level': 'error', 'forbid': False, 'min-spaces-inside': 0, 'max-spaces-inside': 0, 'min-spaces-inside-empty': -1, 'max-spaces-inside-empty': -1}, 'colons': {'level': 'error', 'max-spaces-before': 0, 'max-spaces-after': 1}, 'commas': {'level': 'error', 'max-spaces-before': 0, 'min-spaces-after': 1, 'max-spaces-after': 1}, 'comments': {'level': 'warning', 'require-starting-space': True, 'ignore-shebangs': True, 'min-spaces-from-content': 1}, 'comments-indentation': False, 'document-end': False, 'document-start': False, 'empty-lines': {'level': 'error', 'max': 2, 'max-start': 0, 'max-end': 0}, 'empty-values': False, 'float-values': False, 'hyphens': {'level': 'error', 'max-spaces-after': 1}, 'indentation': {'level': 'error', 'spaces': 'consistent', 'indent-sequences': True, 'check-multi-line-strings': False}, 'key-duplicates': {'level': 'error'}, 'key-ordering': False, 'line-length': {'level': 'error', 'max': 160, 'allow-non-breakable-words': True, 'allow-non-breakable-inline-mappings': False}, 'new-line-at-end-of-file': {'level': 'error'}, 'new-lines': {'level': 'error', 'type': 'unix'}, 'octal-values': {'forbid-implicit-octal': True, 'forbid-explicit-octal': True, 'level': 'error'}, 'quoted-strings': False, 'trailing-spaces': {'level': 'error'}, 'truthy': {'level': 'warning', 'allowed-values': ['true', 'false'], 'check-keys': True}}
INFO     Running ansible-galaxy collection install -v -r collections/requirements.yml -p /Users/jholland/.cache/ansible-compat/da1f3a/collections
INFO     Set ANSIBLE_LIBRARY=/Users/jholland/.cache/ansible-compat/da1f3a/modules:/Users/jholland/.ansible/plugins/modules:/usr/share/ansible/plugins/modules
INFO     Set ANSIBLE_COLLECTIONS_PATH=/Users/jholland/.cache/ansible-compat/da1f3a/collections:/Users/jholland/.ansible/collections:/usr/share/ansible/collections
INFO     Set ANSIBLE_ROLES_PATH=/Users/jholland/.cache/ansible-compat/da1f3a/roles:/Users/jholland/.ansible/roles:/usr/share/ansible/roles:/etc/ansible/roles
INFO     Executing syntax check on playbook upgrade_single.yml (1.21s)
DEBUG    Examining upgrade_single.yml of type playbook
DEBUG    NAME 'backup_config' (prefix='')
DEBUG    VBAR '|' (prefix='')
DEBUG    NAME 'bool' (prefix='')
DEBUG    NEWLINE '\n' (prefix=' ')
DEBUG    ENDMARKER '' (prefix='')
DEBUG    Stop.
DEBUG    NAME 'backup_config' (prefix='')
DEBUG    VBAR '|' (prefix=' ')
DEBUG    NAME 'bool' (prefix=' ')
DEBUG    NEWLINE '\n' (prefix='')
DEBUG    ENDMARKER '' (prefix='')
DEBUG    Stop.
DEBUG    NAME 'device' (prefix='')
DEBUG    NEWLINE '\n' (prefix=' ')
DEBUG    ENDMARKER '' (prefix='')
DEBUG    Stop.
DEBUG    NAME 'device' (prefix='')
DEBUG    NEWLINE '\n' (prefix='')
DEBUG    ENDMARKER '' (prefix='')
DEBUG    Stop.
DEBUG    NAME 'backup_filename' (prefix='')
DEBUG    NEWLINE '\n' (prefix=' ')
DEBUG    ENDMARKER '' (prefix='')
DEBUG    Stop.
DEBUG    NAME 'backup_filename' (prefix='')
DEBUG    NEWLINE '\n' (prefix='')
DEBUG    ENDMARKER '' (prefix='')
DEBUG    Stop.
DEBUG    NAME 'device' (prefix='')
DEBUG    NEWLINE '\n' (prefix=' ')
DEBUG    ENDMARKER '' (prefix='')
DEBUG    Stop.
DEBUG    NAME 'device' (prefix='')
DEBUG    NEWLINE '\n' (prefix='')
DEBUG    ENDMARKER '' (prefix='')
DEBUG    Stop.
DEBUG    NAME 'version' (prefix='')
DEBUG    NEWLINE '\n' (prefix=' ')
DEBUG    ENDMARKER '' (prefix='')
DEBUG    Stop.
DEBUG    NAME 'version' (prefix='')
DEBUG    NEWLINE '\n' (prefix='')
DEBUG    ENDMARKER '' (prefix='')
DEBUG    Stop.
DEBUG    NAME 'result' (prefix='')
DEBUG    NAME 'is' (prefix=' ')
DEBUG    NAME 'not' (prefix=' ')
DEBUG    NAME 'failed' (prefix=' ')
DEBUG    NAME 'and' (prefix=' ')
DEBUG    LPAR '(' (prefix=' ')
DEBUG    NAME 'result' (prefix='')
DEBUG    DOT '.' (prefix='')
DEBUG    NAME 'stdout' (prefix='')
DEBUG    VBAR '|' (prefix=' ')
DEBUG    NAME 'from_json' (prefix=' ')
DEBUG    RPAR ')' (prefix='')
DEBUG    DOT '.' (prefix='')
DEBUG    NAME 'response' (prefix='')
DEBUG    DOT '.' (prefix='')
DEBUG    NAME 'result' (prefix='')
DEBUG    EQEQUAL '==' (prefix=' ')
DEBUG    STRING "'yes'" (prefix=' ')
DEBUG    NEWLINE '\n' (prefix=' ')
DEBUG    ENDMARKER '' (prefix='')
DEBUG    Stop.
DEBUG    NAME 'result' (prefix='')
DEBUG    NAME 'is' (prefix=' ')
DEBUG    NAME 'not' (prefix=' ')
DEBUG    NAME 'failed' (prefix=' ')
DEBUG    NAME 'and' (prefix=' ')
DEBUG    LPAR '(' (prefix=' ')
DEBUG    NAME 'result' (prefix='')
DEBUG    DOT '.' (prefix='')
DEBUG    NAME 'stdout' (prefix='')
DEBUG    VBAR '|' (prefix=' ')
DEBUG    NAME 'from_json' (prefix=' ')
DEBUG    RPAR ')' (prefix='')
DEBUG    DOT '.' (prefix='')
DEBUG    NAME 'response' (prefix='')
DEBUG    DOT '.' (prefix='')
DEBUG    NAME 'result' (prefix='')
DEBUG    EQEQUAL '==' (prefix=' ')
DEBUG    STRING "'yes'" (prefix=' ')
DEBUG    NEWLINE '\n' (prefix='')
DEBUG    ENDMARKER '' (prefix='')
DEBUG    Stop.
DEBUG    NAME 'device' (prefix='')
DEBUG    NEWLINE '\n' (prefix=' ')
DEBUG    ENDMARKER '' (prefix='')
DEBUG    Stop.
DEBUG    NAME 'device' (prefix='')
DEBUG    NEWLINE '\n' (prefix='')
DEBUG    ENDMARKER '' (prefix='')
DEBUG    Stop.
WARNING  Listing 5 violation(s) that are fatal
upgrade_single.yml:36: name[play]: All plays should be named.
upgrade_single.yml:53: args[module]: unable to evaluate string as dictionary (warning)
upgrade_single.yml:53: jinja[spacing]: Jinja2 spacing could be improved: backup_config|bool -> backup_config | bool (warning)
upgrade_single.yml:59: args[module]: unable to evaluate string as dictionary (warning)
upgrade_single.yml:69: args[module]: unable to evaluate string as dictionary (warning)
DEBUG    Attempting to release lock 4339210160 on /Users/jholland/.cache/ansible-compat/da1f3a/.lock
DEBUG    Lock 4339210160 released on /Users/jholland/.cache/ansible-compat/da1f3a/.lock
Read documentation for instructions on how to ignore specific rule violations.

DEBUG    Determined rule-profile order: {'internal-error': (0, 'min'), 'load-failure': (1, 'min'), 'parser-error': (2, 'min'), 'syntax-check': (3, 'min'), 'command-instead-of-module': (4, 'basic'), 'command-instead-of-shell': (5, 'basic'), 'deprecated-bare-vars': (6, 'basic'), 'deprecated-command-syntax': (7, 'basic'), 'deprecated-local-action': (8, 'basic'), 'deprecated-module': (9, 'basic'), 'inline-env-var': (10, 'basic'), 'key-order': (11, 'basic'), 'literal-compare': (12, 'basic'), 'jinja': (13, 'basic'), 'no-jinja-when': (14, 'basic'), 'no-tabs': (15, 'basic'), 'partial-become': (16, 'basic'), 'playbook-extension': (17, 'basic'), 'role-name': (18, 'basic'), 'schema': (19, 'basic'), 'name': (20, 'basic'), 'var-naming': (21, 'basic'), 'yaml': (22, 'basic'), 'name[template]': (23, 'moderate'), 'name[imperative]': (24, 'moderate'), 'name[casing]': (25, 'moderate'), 'no-free-form': (26, 'moderate'), 'spell-var-name': (27, 'moderate'), 'avoid-implicit': (28, 'safety'), 'latest': (29, 'safety'), 'package-latest': (30, 'safety'), 'risky-file-permissions': (31, 'safety'), 'risky-octal': (32, 'safety'), 'risky-shell-pipe': (33, 'safety'), 'galaxy': (34, 'shared'), 'ignore-errors': (35, 'shared'), 'layout': (36, 'shared'), 'meta-incorrect': (37, 'shared'), 'meta-no-info': (38, 'shared'), 'meta-no-tags': (39, 'shared'), 'meta-video-links': (40, 'shared'), 'meta-version': (41, 'shared'), 'meta-runtime': (42, 'shared'), 'no-changed-when': (43, 'shared'), 'no-changelog': (44, 'shared'), 'no-handler': (45, 'shared'), 'no-relative-paths': (46, 'shared'), 'max-block-depth': (47, 'shared'), 'max-tasks': (48, 'shared'), 'unsafe-loop': (49, 'shared'), 'avoid-dot-notation': (50, 'production'), 'disallowed-ignore': (51, 'production'), 'fqcn': (52, 'production'), 'import-task-no-when': (53, 'production'), 'meta-no-dependencies': (54, 'production'), 'single-entry-point': (55, 'production'), 'use-loop': (56, 'production')}
                   Rule Violation Summary                    
 count tag            profile rule associated tags           
     1 jinja[spacing] basic   formatting (warning)           
     1 name[play]     basic   idiom                          
     3 args[module]           syntax, experimental (warning) 

Failed after min profile: 1 failure(s), 4 warning(s) on 1 files.
```
Please give some details of what is happening.
Include a [minimum complete verifiable example] with:

- minimized playbook to reproduce the error
  - Created [this playbook](https://github.com/jamesholland-uk/playground/blob/main/pan-os-ansible/mcve.yml) as the MCVE
- the output of running ansible-lint including the command line used
  - Already mentioned above
- if you're getting a stack trace, also the output of
  `ansible-playbook --syntax-check playbook`
  - N/A (no stack trace is produced)

**Comment by ssbarnea** [comment]
The bug is valid and I already have some code written but i would very useful if you could find a way to reproduce **the same error message** with one of **ansible.builtin** modules. I tried myself to spot few ones that accept "dict" for one of the arguments but I failed to reproduce the same message, I go something a little bit different but exact enough.

**Comment by nre-ableton** [comment]
@ssbarnea I can definitely reproduce it with `ansible.builtin.service`. I'm pretty sure this is a regression from 6.13.1. We removed our `noqa` suppressions on such tasks once 6.13.1 is fixed, but now they are raising `args[module]` errors in 6.14.0.
